src/natcap/invest/managed_raster/managed_raster.pyx

```diff
@@ -300,3 +310,118 @@ cdef class _ManagedRaster:
         if self.write_mode:
             raster_band = None
             raster = None
+
+
+cdef class ManagedFlowDirRaster(_ManagedRaster):
+
+    cdef bint is_local_high_point(self, long xi, long yi):
+        """Check if a given pixel is a local high point.
+
+        Args:
+            xi (int): x coordinate of the pixel to consider
+            yi (int): y coordinate of the pixel to consider
+
+        Returns:
+            True if the pixel is a local high point, i.e. it has no
+            upslope neighbors; False otherwise.
+        """
+        ns = list(self.yield_upslope_neighbors(xi, yi))
+        if ns:
+            return False
+        return True
+
+    @cython.cdivision(True)
+    cdef vector[NeighborTuple] yield_upslope_neighbors(
+            ManagedFlowDirRaster self, long xi, long yi):
+        """Yield upslope neighbors of a given pixel.
+
+        Args:
+            xi (int): x coordinate of the pixel i to consider
+            yi (int): y coordinate of the pixel i to consider
+
+        Yields:
+            (j, xj, yj, p_ij) tuples representing neighboring pixels, where
+            j is the index 0-7 of the neighbor,
+            xj is the x coordinate of pixel j in pixel space,
+            yj is the y coordinate of pixel j in pixel space, and
+            p_ij is the proportion of flow from pixel j that flows into pixel i
+        """
+        cdef int n_dir, flow_dir_j, idx
+        cdef long xj, yj
+        cdef float flow_ji, flow_dir_j_sum
+
+        cdef NeighborTuple n
+        cdef vector[NeighborTuple] upslope_neighbor_tuples
+
+        for n_dir in xrange(8):
```

Comment:
Is `xrange` valid in Cython still? I know it was deprecated in Python 3.

Reply:
Cython still supports it, we use it a few times farther up in `ManagedRaster`. I updated these to `range`.